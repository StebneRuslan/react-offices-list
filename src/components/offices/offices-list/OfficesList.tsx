import React, { Component } from 'react';
import { OfficesListWrapper, OfficeControlsContainer, OfficeCount, OfficeCardsContainer } from './elements';
import { OfficesTitle, OfficesTitleWrapper } from '../elements';
import { Description } from '../../shared/Description';
import { OfficesFooter } from '../offices-footer/OfficesFooter';
import { Button } from '../../shared/Button';
import { OfficeCard } from '../office-card/OfficeCard';
import { OfficeForm } from '../office-form/OfficeForm';
import { OfficeModel } from '../../../models/office.model';

export class OfficesList extends Component<any, any> {
  constructor(props: any) {
    super(props);
    this.state = {
      showForm: false,
      offices: []
    };
  }
  
  componentDidMount(): void {
    this.props.getOfficeList();
  }
  
  handleFormState() {
    this.setState({showForm: !this.state.showForm });
  }
  
  addOffice(office: OfficeModel) {
    this.props.createOffice(office);
    this.setState({ showForm: false })
  }
  
  updateOffice(newOffice: OfficeModel): void {
    const position = this.state.offices.findIndex((office: any) => office.name === newOffice.name);
    const editableOffice = Object.assign({}, { ...newOffice });
    if (position > -1) {
      this.setState({
        offices: Object.assign([], this.state.offices, { [position]: editableOffice })
      })
    }
  }
  
  renderNewForm(): any {
    return this.state.showForm
      ? <OfficeForm
        saveOffice={this.addOffice.bind(this)}
        closeForm={this.handleFormState.bind(this)}
      />
      : null;
  }
  
  renderEditableForm(office: OfficeModel, index: number): any {
    return <OfficeForm
      key={index}
      editable={true}
      saveOffice={this.updateOffice.bind(this)}
      closeForm={this.handleEditForm.bind(this)}
      office={office}
    />
  }
  
  removeOffice(name: string): void {
    this.props.removeOffice(name);
  }
  
  handleEditForm(name: string, editable: boolean): void {
    const position = this.state.offices.findIndex((office: any) => office.name === name);
    const editableOffice = Object.assign({}, { ...this.state.offices[position], editable });
    if (position > -1) {
      this.setState({ offices: Object.assign([], this.state.offices, { [position]: editableOffice }) })
    }
  }
  
  render() {
    return (
      <OfficesListWrapper>
        <OfficesTitleWrapper>
          <OfficesTitle>OFFICES</OfficesTitle>
          <OfficesTitle disabled>COMPANY INFO</OfficesTitle>
        </OfficesTitleWrapper>
        <Description>Updating your location and contact information helps you appeal to regional investors and service providers.</Description>
        <OfficeCardsContainer>
          <OfficeControlsContainer>
            <Button white onClick={this.handleFormState.bind(this)}>Add New Office</Button>
            <OfficeCount>{this.state.offices.length} Offices</OfficeCount>
          </OfficeControlsContainer>
          {this.renderNewForm()}
          {this.props.offices.map((office: any, index: number) => {
            return office.editable
              ? this.renderEditableForm(office, index)
              : <OfficeCard
              key={index}
              office={office}
              removeOffice={this.removeOffice.bind(this)}
              editOffice={this.handleEditForm.bind(this)}
            />
          })}
        </OfficeCardsContainer>
        <OfficesFooter />
      </OfficesListWrapper>
    )
  }
}
