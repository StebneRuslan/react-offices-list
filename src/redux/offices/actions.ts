import actionTypes from './action-types'
import { OfficeModel } from '../../models/office.model';

const toggleNewOfficeForm = () => ({ type: actionTypes.TOGGLE_NEW_OFFICE_FORM });
const toggleEditableOfficeForm = (data: string) => ({ type: actionTypes.TOGGLE_EDITABLE_OFFICE_FORM, payload: data });

const getOfficeList = () => ({ type: actionTypes.GET_OFFICES });
const getOfficeListSuccess = (data: OfficeModel[]) => ({ type: actionTypes.GET_OFFICES_SUCCESS, payload: data });
const getOfficeListError = (data: string) => ({ type: actionTypes.GET_OFFICES_ERROR, err: data });

const createOffice = (data: OfficeModel) => ({ type: actionTypes.CREATE_OFFICE, payload: data});
const createOfficeSuccess = (data: OfficeModel) => ({ type: actionTypes.CREATE_OFFICE_SUCCESS, payload: data});
const createOfficeError = (data: string) => ({ type: actionTypes.CREATE_OFFICE_ERROR, err: data});

const removeOffice = (data: string) => ({ type: actionTypes.REMOVE_OFFICE, payload: data});
const removeOfficeSuccess = (data: string) => ({ type: actionTypes.REMOVE_OFFICE_SUCCESS, payload: data});
const removeOfficeError = (data: string) => ({ type: actionTypes.REMOVE_OFFICE_ERROR, err: data});

const updateOffice = (data: OfficeModel) => ({ type: actionTypes.UPDATE_OFFICE, payload: data});
const updateOfficeSuccess = (data: OfficeModel) => ({ type: actionTypes.UPDATE_OFFICE_SUCCESS, payload: data});
const updateOfficeError = (data: string) => ({ type: actionTypes.UPDATE_OFFICE_ERROR, payload: data});

export default {
  toggleNewOfficeForm,
  toggleEditableOfficeForm,
  getOfficeList,
  getOfficeListSuccess,
  getOfficeListError,
  createOffice,
  createOfficeSuccess,
  createOfficeError,
  removeOffice,
  removeOfficeSuccess,
  removeOfficeError,
  updateOffice,
  updateOfficeSuccess,
  updateOfficeError
}
