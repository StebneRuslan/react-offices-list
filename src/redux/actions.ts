import offices from './offices/actions'

export default {
  offices
}
