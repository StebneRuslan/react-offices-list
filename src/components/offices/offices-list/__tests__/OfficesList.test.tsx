import React from 'react';
import { shallow } from 'enzyme';
import { OfficesList, OfficesListInterface } from '../OfficesList';
import actions from '../../../../redux/offices/actions';

const props: OfficesListInterface = {
  isFetching: false,
  toggleNewOfficeForm: jest.fn(),
  getOfficeList: () => jest.fn().mockReturnValue(actions.getOfficeList()),
  offices: [{
    officeType: false,
    name: 'name',
    editable: false,
    city: 'city',
    country: 'country',
    email: 'email@appcreative.net',
    fax: 'fax',
    phone: '88888888',
    postalCode: 'code',
    secondAddress: 'secondAddress',
    state: 'state',
    streetAddress: 'secondAddress'
  }],
  showForm: false,
};

describe('<OfficesList />', () => {
  it('OfficeCard renders without crashing', () => {
    const component = shallow(<OfficesList { ...props } />);
    expect(component).toMatchSnapshot()
  });
  
  it('Should call the toggleNewOfficeForm function when clicked', () => {
    const wrapper = shallow(<OfficesList { ...props } />);
    expect(props.toggleNewOfficeForm).toHaveBeenCalledTimes(0);
    wrapper.find('#toggle-new-office-form').at(0).simulate('click');
    expect(props.toggleNewOfficeForm).toHaveBeenCalledTimes(1);
  });
});
