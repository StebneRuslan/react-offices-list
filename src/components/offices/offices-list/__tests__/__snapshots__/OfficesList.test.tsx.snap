// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OfficesList /> OfficeCard renders without crashing 1`] = `
<styled.div>
  <styled.div>
    <styled.h1>
      OFFICES
    </styled.h1>
    <styled.h1
      disabled={true}
    >
      COMPANY INFO
    </styled.h1>
  </styled.div>
  <styled.div>
    Updating your location and contact information helps you appeal to regional investors and service providers.
  </styled.div>
  <styled.div>
    <styled.div>
      <styled.button
        id="toggle-new-office-form"
        onClick={[MockFunction]}
        white={true}
      >
        Add New Office
      </styled.button>
      <styled.p>
        1
         Offices
      </styled.p>
    </styled.div>
    <Connect(OfficeCard)
      key="0"
      office={
        Object {
          "city": "city",
          "country": "country",
          "editable": false,
          "email": "email@appcreative.net",
          "fax": "fax",
          "name": "name",
          "officeType": false,
          "phone": "88888888",
          "postalCode": "code",
          "secondAddress": "secondAddress",
          "state": "state",
          "streetAddress": "secondAddress",
        }
      }
    />
  </styled.div>
  <OfficesFooter />
</styled.div>
`;
